packages/smart-queue/src/enqueue.js

```diff
@@ -0,0 +1,45 @@
+// @flow
+import defaults from './defaults';
+import {
+  getQueueFunction,
+  getQueueMeta,
+  indexOfAction,
+  methodAllowed,
+  safeToProceed,
+  valid
+} from './utils';
+import type { Config, Outbox, OfflineAction, Context } from './types';
+
+const WARNING =
+  'offline.queue must be an object which contains a valid "method" and a unique "key"';
+
+export const enqueue = (userConfig: $Shape<Config> = {}) => {
+  const config = { ...defaults, ...userConfig };
+
+  return (outbox: Outbox, action: OfflineAction, context: Context): Outbox => {
+    const queueMeta = getQueueMeta(action);
+
+    if (!queueMeta) {
+      return [...outbox, action];
+    }
+
+    if (!valid(queueMeta)) {
+      // eslint-disable-next-line no-console
+      console.error(new Error(WARNING));
+      return outbox;
+    }
+
+    const queueFunction = getQueueFunction(queueMeta);
+    const index = indexOfAction(outbox, queueMeta);
```

Comment:
`indexOfAction` should also contain the check for `methodAllowed` as certain actions cannot be squashed over each other. We should aim to find the best available, not give up on the first id match which may not allow a merge strategy to take place. E.g. `enqueue([Read, Update], Update) -> [Read, Update, Update]`, first check fails so we add to the queue as new entity, where what we actually want is `[Read, Update]`